x/exchange/keeper/market.go

```diff
@@ -0,0 +1,1303 @@
+func (k Keeper) CanCreateBid(ctx sdk.Context, marketID uint32, addr sdk.AccAddress) bool {
+	reqAttrs := k.GetReqAttrsBid(ctx, marketID)
+	return k.acctHasReqAttrs(ctx, addr, reqAttrs)
+}
+
+// UpdateReqAttrs updates the required attributes in the store using the provided changes.
+// The caller is responsible for making sure this update should be allowed (e.g. by calling CanManageReqAttrs first).
+func (k Keeper) UpdateReqAttrs(ctx sdk.Context, msg *exchange.MsgMarketManageReqAttrsRequest) error {
+	admin := sdk.MustAccAddressFromBech32(msg.Admin)
+
+	var errs []error
+	// We don't care if the attributes to remove are valid so that we
+	// can remove entries that are somehow now invalid.
+	askToRemove, _ := exchange.NormalizeReqAttrs(msg.CreateAskToRemove)
+	askToAdd, err := exchange.NormalizeReqAttrs(msg.CreateAskToAdd)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	bidToRemove, _ := exchange.NormalizeReqAttrs(msg.CreateBidToRemove)
+	bidToAdd, err := exchange.NormalizeReqAttrs(msg.CreateBidToAdd)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	marketID := msg.MarketId
+	store := k.getStore(ctx)
+
+	if err = updateReqAttrsAsk(store, marketID, askToRemove, askToAdd); err != nil {
+		errs = append(errs, err)
+	}
+	if err = updateReqAttrsBid(store, marketID, bidToRemove, bidToAdd); err != nil {
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	k.emitEvent(ctx, exchange.NewEventMarketReqAttrUpdated(marketID, admin))
+	return nil
+}
+
+// getMarketAccountByAddr gets a market's account given its address.
+// This is for when you've already called exchange.GetMarketAddress(marketID) and need it for other things too.
+func (k Keeper) getMarketAccountByAddr(ctx sdk.Context, marketAddr sdk.AccAddress) *exchange.MarketAccount {
+	acc := k.accountKeeper.GetAccount(ctx, marketAddr)
+	if acc == nil {
+		return nil
+	}
+	marketAcc, ok := acc.(*exchange.MarketAccount)
+	if !ok {
+		return nil
+	}
+	return marketAcc
+}
+
+// GetMarketAccount gets a market's account from the account module.
+func (k Keeper) GetMarketAccount(ctx sdk.Context, marketID uint32) *exchange.MarketAccount {
+	marketAddr := exchange.GetMarketAddress(marketID)
+	return k.getMarketAccountByAddr(ctx, marketAddr)
+}
+
+// GetMarketDetails gets a market's details.
+func (k Keeper) GetMarketDetails(ctx sdk.Context, marketID uint32) *exchange.MarketDetails {
+	marketAcc := k.GetMarketAccount(ctx, marketID)
+	if marketAcc == nil {
+		return nil
+	}
+	return &marketAcc.MarketDetails
+}
+
+// UpdateMarketDetails updates a market's details. It returns an error if the market account
+// isn't found or if there aren't any changes provided.
+func (k Keeper) UpdateMarketDetails(ctx sdk.Context, marketID uint32, marketDetails exchange.MarketDetails, updatedBy sdk.AccAddress) error {
+	if err := marketDetails.Validate(); err != nil {
```

Comment:
It looked like this was already validated in the ValidateBasic. Does this need to be done twice?

Reply:
This is a public keeper function. So if someone is trying to do things with the exchange in some other module/place, they might be using this, in which case, we do need to make sure the market details are valid before we save them. Otherwise, someone might accidentally (or purposefully) put a giant amount of data in there and cause problems.

Honestly, I'm on the fence about most of the public keeper functions for this module. I could probably clean up a lot of stuff if we didn't want to support other modules that want to interact with the exchange. I'm pretty sure there's several public keeper funcs (e.g. `GetAccessGrants`) that aren't called anywhere, but exist because we usually have such public funcs in keepers in case something else needs access to that info.

Once I dig into the keeper unit tests, I'll have a better grasp of how I feel about those.